tools/create_db.py

```diff
@@ -0,0 +1,153 @@
+        if not pt:
+            logging.warning(f"No patch for: {train}")
+        elif len(pt) > 1:
+            logging.warning(f"can't match patch from {list(pt.keys())} patches for: {train}")
+        else:
+            ifs, ifs_sha, ifs_patch, offset, ifs_checksum = list(pt.values())[0]
+            detail["SHA1_patch"] = ifs_sha
+            detail["ifs_header_checksum_patch"] = ifs_patch
+            detail["RCC_offset"] = offset
+            detail["ifs_file_name_checksum_patch"] = ifs_checksum
+            #print(ifs_patch)
+        
+        detail["PN_model"], detail["PN_INDEX"], detail["PN_Ident"] = PN1_split(detail["PN1"])
+        
+        row = FIELD_SEP.join([b2s(detail.get(f, "")) for f in HEADINGS])
+        csv.append(row)
+        
+    logging.info(f"database written: {db_file}")
+    db_file.write_text("\n".join(csv))
+
+
+def b2s(b):
+    return str(b, "utf8") if isinstance(b, bytes) else str(b)
+
+
+def train_split(train):
+    hdr, brand, region, *vers = train.upper().split("_")
+    return region, brand
+
+def ifs_name_split(file_name):
+    MU, ifs1, ifs2, ifs3, offset, ifs_checksum = file_name.lower().split("-")
+    return offset, ifs_checksum
+
+def PN1_split(pn1):
+    PN_model = pn1[0:6]
+    PN_Ident = pn1[6:9]
+    try:
+        PN_INDEX = pn1[9]
+    except IndexError:
+        PN_INDEX = ''
+    return PN_model, PN_INDEX, PN_Ident
+
+
+def sha1sum(filename):
+    h  = hashlib.sha1()
+    b  = bytearray(128*1024)
+    mv = memoryview(b)
+    with open(filename, 'rb', buffering=0) as f:
+        while n := f.readinto(mv):
+            h.update(mv[:n])
+    return h.hexdigest()
+
+
+def parse_backups(backups_dir: Path):
+    logging.info(f"Parsing backups in: {backups_dir} ({backups_dir.exists()})")
+    backup_details = list()
+    for eeprom in backups_dir.glob("*/*-EEProm.bin"): # Change to "**/*-EEProm.bin" to enable scan of sub directories
+        backup = eeprom.parent
+        logging.info(f"Backup: {backup.name}")
+        if os.path.getsize(eeprom) != 8192:
+            logging.warning(f"Unexpected eeprom size: {os.path.getsize(eeprom)}")
+        with eeprom.open("rb") as data:
+            details = parse_eeprom(data)
+        for ifs in backup.glob("*-ifs-root-part2*.ifs"):
+            details["ifs_SHA1"] = sha1sum(ifs)
+        for ifs_header in backup.glob("*-ifs-root-part2*.ifs"):
+            with ifs_header.open("rb") as data:
+                ifs_details = parse_ifs(data)
+                details.update(ifs_details)
+
+        backup_details.append(details)
+
+    return backup_details
+
+def parse_patches(patches_dir: Path):
+    logging.info("Scanning patches")
+    patch_details = dict()
+    ifs_patch_details = list()
+    for ifs in patches_dir.glob("*/*.ifs"):
+        patches = ifs.parent
+        train = ifs.parent.name.replace("_PATCH", "")
+        logging.warning(f"{train}: {ifs}")
+        sha1 = sha1sum(ifs)
+        for ifs_header in patches.glob("*-ifs-root-part2*.ifs"):
```

Comment:
@harman-f  This loop to "search and process all ifs part2 files" is _inside_ the "search all ifs files" above on line 119, so week be re-run for every ifs file.

Reply:
Done